### PR TITLE
Fixing body streamers

### LINKED ISSUE #50: Preflight OPTIONS requests fail with 408 - Request timeout
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

When using `supabase functions serve` to serve all functions simultaneously, the pre-flight 'OPTIONS' request initiated by a browser agent fails.
The server returns Status Code `408 - Request timeout` with a message body of `{"msg":"Request could not be processed by the server because it timed out or an error was thrown."}`

**The bug is triggered by the fact that the OPTIONS body is empty**

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Use code snippet at [https://supabase.com/docs/guides/functions/cors#recommended-setup](https://supabase.com/docs/guides/functions/cors#recommended-setup) to setup CORS headers in the `serve` function
2. Send `OPTIONS` request to edge function

## Expected behavior

The server should return `200` with the CORS headers set appropriately.

## Screenshots

Capture d’écran 2023-04-13 à 15 22 38
Capture d’écran 2023-04-13 à 15 23 28

## System information

- OS: macOS
- Browser (if applies) : chrome. Also tested outside browser (Postman)
- Version of supabase-js: 2.20.0
- Version of Node.js: not relevant

## Additional context

1. If serving the function individually via `supabase functions serve <function-name>`, the bug does not happen. It only happens when serving all functions simultaneously via `supabase functions serve`

2. If calling the function with the `POST` method directly, the bug does not happen. 

3. Inserting `console.log()` statements everywhere in the edge function code shows that the script reaches the point where it has to send the response to the OPTIONS request, then crashes silently.

4. If you put some dummy data into the body of the `OPTIONS` request, the bug does not happen:
Capture d’écran 2023-04-13 à 15 24 00

*Unfortunately a browser-generated pre-flight request always has an empty body*

To me it looks like when I call `return new Response('ok', { headers: corsHeaders })` inside the edge function, this triggers the bug because probably the server-side is still waiting for the client to send its body and it does not want to reply with the Response before it can close the incoming connection - but I may be mistaken on my interpretation.

**In conclusion, it looks like there may be some code inside edge-runtime that tries to enforce a non-empty body on an `OPTIONS` request ?**

### PR DESCRIPTION
**Pull request by andreespirela**
## What kind of change does this PR introduce?

Fixes #50 .

This fix corrects the way we identify when a request has a body.

### CODE REVIEW

File: crates/sb_workers/user_workers.js [hunk 1]
```diff
@@ -18,16 +18,29 @@ const ops = core.ops;
...
 class UserWorker {
     constructor(key) {
         this.key = key;
     }
 
     async fetch(req) {
-        const { method, url, headers, body, bodyUsed } = req;
+        const { method, url, headers, body } = req;
+
+        const bodyData = (await new Response(body).arrayBuffer());
```
Comment (andreespirela) [inline crates/sb_workers/user_workers.js:39]: This fix is a little odd. cc @laktek Let's get some eyes here.

**Comment by vejja** [comment]
@andreespirela @laktek : sorry to intervene in a field where I certainly don't get all the complexities, but I do feel that `hasBody` is a bit hacky here. Especially relying on parsing the headers and looking up for `content-length` which is not guaranteed to be correct.

Why not rely on the standard [`fetch`](https://fetch.spec.whatwg.org/#ref-for-dom-body-bodyused) API spec? 
It says that:
> The body getter steps are to return null if [this](https://webidl.spec.whatwg.org/#this)’s [body](https://fetch.spec.whatwg.org/#concept-body-body) is null; otherwise [this](https://webidl.spec.whatwg.org/#this)’s [body](https://fetch.spec.whatwg.org/#concept-body-body)’s [stream](https://fetch.spec.whatwg.org/#concept-body-stream).
>
> The bodyUsed getter steps are to return true if [this](https://webidl.spec.whatwg.org/#this)’s [body](https://fetch.spec.whatwg.org/#concept-body-body) is non-null and [this](https://webidl.spec.whatwg.org/#this)’s [body](https://fetch.spec.whatwg.org/#concept-body-body)’s [stream](https://fetch.spec.whatwg.org/#concept-body-stream) is [disturbed](https://streams.spec.whatwg.org/#is-readable-stream-disturbed); otherwise false.

In summary, `hasBody` should be equal to

- `body` is not null
- AND : `bodyUsed` is false

Am I mistaken here?

File: crates/sb_workers/user_workers.js [hunk 2]
```diff
@@ -36,11 +36,8 @@ class UserWorker {
     async fetch(req) {
         const { method, url, headers, body } = req;
```
Comment (vejja) [inline crates/sb_workers/user_workers.js:27]: const { method, url, headers, body, bodyUsed } = req;

File: crates/sb_workers/user_workers.js [hunk 3]
```diff
@@ -36,11 +36,8 @@ class UserWorker {
...
 
-        const bodyData = (await new Response(body).arrayBuffer());
-        const bodyLength = bodyData.byteLength;
-
         const headersArray = Array.from(headers.entries());
-        const hasBody = bodyLength > 0;
+        const hasBody = headersArray.some(([headerName, headerValue]) => headerName.toLowerCase() === 'content-length' && Number(headerValue) > 0);
```
Comment (vejja) [inline crates/sb_workers/user_workers.js:30]: const hasBody = (body !== null) && (!bodyUsed);

**Review by vejja — CHANGES_REQUESTED** [review]

File: crates/sb_workers/user_workers.js (hunk 3 above)
Comment (andreespirela) [reply crates/sb_workers/user_workers.js:30]: I will add the `body !== null`. `!(bodyUsed)` is unnecessary, but checking if body is null is okay. Note that for the time being, `body` will always have a value. It's supposed to be null when it's empty, but the tech behind this is not respecting the spec. It's an upstream

Comment (vejja) [reply crates/sb_workers/user_workers.js]: The reason for checking !bodyUsed is that if the body stream has already been read, you are piping a closed stream to the Rust worker, which I would imagine could trigger all sorts of errors

Comment (vejja) [reply crates/sb_workers/user_workers.js]: > `body` will always have a value. It's supposed to be null when it's empty, but the tech behind this is not respecting the spec. It's an upstream

Where does the request come from ? Isn't it a JS object coming from Deno's std/http/server ?

**Review by vejja — CHANGES_REQUESTED** [review]

File: crates/sb_workers/user_workers.js (hunk 3 above)
Comment (andreespirela) [reply crates/sb_workers/user_workers.js]: At the time of using `bodyUsed` it's never been read. In fact we need to figure out whether `hasBody` is true or false before we even pipe it. If we for some reason consume the body and then pipe it, tests and everything will fail. Which is why it's designed -> Figure out whether it has a body _without consuming it_ -> Pipe it.

File: crates/sb_workers/user_workers.js [hunk 4]
```diff
@@ -27,7 +27,7 @@ class UserWorker {
         const { method, url, headers, body } = req;
 
         const headersArray = Array.from(headers.entries());
-        const hasBody = headersArray.some(([headerName, headerValue]) => headerName.toLowerCase() === 'content-length' && Number(headerValue) > 0);
+        const hasBody = headersArray.some(([headerName, headerValue]) => headerName.toLowerCase() === 'content-length' && Number(headerValue) > 0) || body == null;
```
[the remark this answers is not in the document]
Comment (andreespirela) [reply crates/sb_workers/user_workers.js:30]: Yes, pushing change

**Comment by vejja** [comment]
An alternative way of doing it
- if you can't rely on `body !== null` (because it always has a value)
- and if you are 100% sure that the stream has never been read at this point

would be to read the body stream at this stage, and then based on whether or not the resulting bytes array is empty, then create a brand new stream that you write down to the worker. 

It's a bit of a shame to force-read a stream to then re-write it, instead of elegantly piping it. But it might be more reliable than relying on `content-length` which the HTTP spec does not make mandatory - and even forbids when `transfer-encoding` is set.

What do you think?

**Comment by vejja** [comment]
> An alternative way of doing it
> 
> * if you can't rely on `body !== null` (because it always has a value)
> * and if you are 100% sure that the stream has never been read at this point
> 
> would be to read the body stream at this stage, and then based on whether or not the resulting bytes array is empty, then create a brand new stream that you write down to the worker.
> 
> It's a bit of a shame to force-read a stream to then re-write it, instead of elegantly piping it. But it might be more reliable than relying on `content-length` which the HTTP spec does not make mandatory - and even forbids when `transfer-encoding` is set.
> 
> What do you think?

Reflecting here on my own suggestions - I think this is silly because it defeats the whole purpose of (1) streams and (2) HTTP chunked transfers... So I'm downvoting myself.

However I gave a quick look at Rust's `hyper` and it looks like if you call the `data()` function on a `Body`, it will return `None` if the body is empty. That might be the safest way to perform the test, however it must be done at the Rust level, not the JS one...

**Comment by laktek** [comment]
Thanks @andreespirela! I opened a PR with a slightly different implementation (#75), which handles the case where request body can be a chunked stream. I will close this in favor of it.
